File: lc_go/lcGo/array/rotate48.go
package main

import "fmt"

func rotate(matrix [][]int) {
    n := len(matrix)

    // Step 1: 转置
    for i := 0; i < n; i++ {
        for j := i + 1; j < n; j++ {
            matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        }
    }

    // Step 2: 每一行反转
    for i := 0; i < n; i++ {
        reverse(matrix[i])
    }
}

func reverse(row []int) {
    left, right := 0, len(row)-1
    for left < right {
        row[left], row[right] = row[right], row[left]
        left++
        right--
    }
}

func main(){
	matrix := [][]int{
		{1,2,3},
		{4,5,6},
		{7,8,9},
	}
	rotate(matrix)
	fmt.Println(matrix)
}

File: lc_java/lcJava/array/rotate48.java
package array;

public class rotate48 {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // Step 1: 转置
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }

        // Step 2: 每一行反转
        for (int i = 0; i < n; i++) {
            reverseRow(matrix[i]);
        }
    }

    private void reverseRow(int[] row) {
        int left = 0, right = row.length - 1;
        while (left < right) {
            int tmp = row[left];
            row[left++] = row[right];
            row[right--] = tmp;
        }
    }

    public static void main(String[] args) {
        rotate48 rotate48 = new rotate48();
        int[][] matrix = {
                {1,2,3},{4,5,6},{7,8,9}
        };
        rotate48.rotate(matrix);
    }
}
